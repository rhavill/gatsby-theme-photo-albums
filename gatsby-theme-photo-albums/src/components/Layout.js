import React from 'react'
import PropTypes from 'prop-types'
import Header from './Header'
import {Layout as ThemeLayout, Main} from 'theme-ui'

const Layout =  ({children, path}) => (
  <ThemeLayout>
    <Header path={path}/>
    <Main>
      {children}
    </Main>
  </ThemeLayout>
)

Layout.propTypes = {
  path: PropTypes.string.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired
}

export default Layout