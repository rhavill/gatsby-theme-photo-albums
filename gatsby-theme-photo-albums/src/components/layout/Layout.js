import React from 'react'
import PropTypes from 'prop-types'
import Header from './Header'
import LayoutStyles from '../../styles/Layout'
import MainContentStyles from '../../styles/MainContent'

const Layout = ({children, path}) => 
  <LayoutStyles>
    <Header path={path}/>
    <MainContentStyles>
      {children}
    </MainContentStyles>
  </LayoutStyles>

Layout.propTypes = {
  path: PropTypes.string.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired
}

export default Layout