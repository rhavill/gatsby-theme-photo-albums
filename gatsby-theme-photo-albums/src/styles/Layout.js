/** @jsx jsx */
import { jsx } from 'theme-ui'
import PropTypes from 'prop-types'

const Layout = ({children}) => 
  <div sx={{variant: 'layout.root'}}>
    {children}
  </div>

Layout.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired
}

export default Layout