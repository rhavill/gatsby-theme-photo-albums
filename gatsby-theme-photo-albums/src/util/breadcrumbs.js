import compose from 'ramda/src/compose'
import {removeFileExtension, toTitleCase} from './url-text'

const breadcrumbs = (path) => {
  // Set path to empty string for root directory. 
  // Also, remove trailing slash (if it exists) for consistency.
  const parts = path === '/' ? [''] : path.replace(/\/$/, '').split('/')
  return parts
    .map((folder, i) => ({
      // convert to "title case" and remove file extension
      title: folder === '' 
        ? '' 
        : compose(toTitleCase, removeFileExtension)(folder),
      path: folder === '' ? '/' : parts.filter((part, j) => j <= i).join('/') 
    }))
}

export default breadcrumbs