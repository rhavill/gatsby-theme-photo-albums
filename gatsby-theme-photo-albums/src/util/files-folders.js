const curry = require('ramda/src/curry')
const filter = require('ramda/src/filter')
const match = require('ramda/src/match')
const {prependbaseUrl} = require('./url-text')

const getChildPaths = curry(
  (currentPath, paths) => {
    return filter(isParentDirectoryOf(currentPath), paths)
  }
)

/**
 * Determines whether a directory is a parent directory of a file or directory.
 * @param {string} parent The path of a possible parent directory. 
 * @param {string} child The path of a file or directory.
 * @returns {bool}
 */
const isParentDirectoryOf = curry(
  (parent, child) => {
    const regex = new RegExp('^' + parent + '/?[^/]+$')
    return match(regex, child).length > 0
  }
)

const getPagerData = (currentPath, files, filesPerPage) => {
  const pagerData = []
  const children = getChildPaths(currentPath, files)
  const childCount = children.length
  const numPages = Math.ceil(childCount / filesPerPage) || 1
  Array.from({ length: numPages }).forEach((_, i) => {
    pagerData.push({
      limit: filesPerPage,
      skip: i * filesPerPage,
      numPages,
      currentPage: i + 1
    })
  })
  return pagerData
}

/**
 * Accepts an array of gatsby-source-filesystem objects that must have a
 * relativePath property. Adds a url property to all of the objects.
 */
const addUrlProps = curry((baseUrl, filesOrDirectories) => 
  filesOrDirectories.map(fileOrDirectory => ({
    ...fileOrDirectory,
    url: prependbaseUrl(baseUrl, fileOrDirectory.relativePath)
  })))

module.exports = {
  addUrlProps,
  getChildPaths,
  getPagerData,
}