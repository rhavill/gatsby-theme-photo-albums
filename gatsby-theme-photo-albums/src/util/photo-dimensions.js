export default (windowDimensions, pageContext) => {
  const {width, height} = windowDimensions
  const ratio = pageContext.width / pageContext.height
  if (pageContext.width < width && pageContext.height < height)
    return {maxWidth: width, maxHeight: height}
  const excessWidthFactor = (pageContext.width - width) / width
  const excessHeightFactor = (pageContext.height - height) / height
  return excessWidthFactor > excessHeightFactor
    ? {maxWidth: width, maxHeight: width / ratio}
    : {maxWidth: height * ratio, maxHeight: height}
}