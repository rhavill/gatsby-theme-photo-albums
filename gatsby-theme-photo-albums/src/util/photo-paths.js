const forEach = require('ramda/src/forEach')
const keys = require('ramda/src/keys')
const match = require('ramda/src/match')
const {indexedForEach, groupByProp} = require('./ramda-utils')
const {removeFileExtension} = require('./url-text')
const groupByRelativeDirectory = groupByProp('relativeDirectory')

const getPhotoPathsWithPages = (photosPerPage, fileData) => {
  const photoPaths = {}
  const groupedFiles = groupByRelativeDirectory(fileData)
  forEach(relativeDirectory => indexedForEach(
    (file, i) => {
      const pageNumber = Math.ceil((i + 1) / photosPerPage)
      photoPaths[file.url] = getPhotoPathWithPage(
        pageNumber, file.url
      )
    }, 
    groupedFiles[relativeDirectory]
  ), keys(groupedFiles))
  return photoPaths
}

const getPhotoPathWithPage = (pageNumber, url) => {
  url = removeFileExtension(url)
  const matches = match(/^(.*\/)([^/]+)$/, url)
  if (matches && matches[1] && matches[2]) {
    const path =  `${matches[1]}` + (pageNumber > 1 ? `${pageNumber}/` : '') + matches[2]
    return path
  }
  return url
}

module.exports = {getPhotoPathsWithPages, getPhotoPathWithPage}