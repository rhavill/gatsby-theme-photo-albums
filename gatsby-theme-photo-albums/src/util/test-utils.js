import mergeDeepRight from 'ramda/src/mergeDeepRight'
import React from 'react'
import PropTypes from 'prop-types'
import { render } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'emotion-theming'

import theme from '../theme'

const create = (element, options) => {
  return renderer.create(
    <ThemeProvider theme={theme}>{element}</ThemeProvider>, options
  )
}

const customRenderer = mergeDeepRight(renderer, {create})

const Wrapper = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      {children}
    </ThemeProvider>
  )
}
Wrapper.propTypes = {
  children: PropTypes.node.isRequired
}

const customRender = (ui, options) =>
  render(ui, { wrapper: Wrapper, ...options })

// re-export everything from React Testing Library
export * from '@testing-library/react'

// override react testing library render method and custom react-test-renderer r
export { customRender as render, customRenderer as renderer }