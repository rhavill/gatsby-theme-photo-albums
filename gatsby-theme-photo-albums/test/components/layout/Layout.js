import React from 'react'
import {cleanup, render, renderer} from '../../../src/util/test-utils'
import Layout from '../../../src/components/layout/Layout'

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup)

describe('Layout', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Layout path='/'><div>hello</div></Layout>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('displays child nodes', () => {
    const {getByText} = render(
      <Layout path='/'><div>hello</div></Layout>
    )
    expect(getByText('hello')).toBeTruthy()
  })
})
