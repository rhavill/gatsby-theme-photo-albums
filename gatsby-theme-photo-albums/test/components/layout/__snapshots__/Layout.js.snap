// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout renders correctly 1`] = `
<div
  className="css-1ala8gf-Layout"
>
  <header
    className="css-a1lwk2-Header"
  >
    <nav>
      <label
        css={
          Object {
            "map": undefined,
            "name": "1ox5w5f",
            "next": undefined,
            "styles": "
    font-size: 2rem;
    padding: .5rem;
    display: inline-block;
    background-color: #6bccf9;
    cursor: pointer;
  ",
            "toString": [Function],
          }
        }
        htmlFor="menu-checkbox"
      >
        â˜°
      </label>
      <input
        css={
          Object {
            "display": "none",
          }
        }
        defaultChecked={false}
        id="menu-checkbox"
        onChange={[Function]}
        type="checkbox"
      />
      <ul
        css={
          Object {
            "map": undefined,
            "name": "1myc36q",
            "next": undefined,
            "styles": "
    list-style-type: none;
    margin: 0rem;
    padding: 0;
    display: none;
    background-color: #6bccf9;
  ",
            "toString": [Function],
          }
        }
      >
        <li
          css={
            Object {
              "map": undefined,
              "name": "lmlfqw",
              "next": undefined,
              "styles": "
    display: block;
    padding: .5rem;
    /*box-sizing: border-box;*/
    border-top: 1px solid;
  ",
              "toString": [Function],
            }
          }
        >
          <a
            href="/"
          >
            <span
              className="first-breadcrumb"
            >
              Home
            </span>
          </a>
        </li>
      </ul>
    </nav>
  </header>
  <main
    className="css-nv2sqw-Main"
  >
    <div>
      hello
    </div>
  </main>
</div>
`;
